mpqp/tools/theoretical_simulation.py

```diff
@@ -0,0 +1,210 @@
+    gives the relation between the two.
+
+    Args:
+        alpha: The distance Jensen-Shannon between the non noisy distribution
+            and the noisy distribution.
+
+    Returns:
+        The trust interval for the distance.
+    """
+    return np.sqrt(alpha)
+
+
+# TODO: this bound it coming out of my ass, see if these results already in
+# statistics. Some useful resources might be: jensen inequality, chernoff bound
+
+
+def trust_int(circuit: QCircuit):
+    """Given a circuit, this computes the trust interval for the output samples
+    given into consideration the noise in the circuit.
+
+    Args:
+        circuit: The circuit.
+
+    Returns:
+        The size of the trust interval (related to the Jensen-Shannon distance).
+    """
+    noiseless_circuit = QCircuit(
+        [inst for inst in circuit.instructions if not isinstance(inst, NoiseModel)]
+    )
+    noiseless_probs = theoretical_probs(noiseless_circuit)
+    noisy_probs = theoretical_probs(circuit)
+    return dist_alpha_matching(float(jensenshannon(noiseless_probs, noisy_probs)))
+
+
+def exp_id_dist(
+    circuit: QCircuit,
+    shots: int = 1024,
+    device: AvailableDevice = AWSDevice.BRAKET_LOCAL_SIMULATOR,
+):
+    """This function computes Jensen-Shannon the distance between the non noisy
+    distribution and the noisy distribution.
+
+    Args:
+        circuit: The circuit.
+        shots: Number of shots in the basis measurement.
+        device: The device to be tested.
+
+    Returns:
+        The distance between the non noisy distribution and the noisy
+        distribution.
+    """
+    noisy_probs = theoretical_probs(circuit)
+
+    noisy_circuit = circuit.without_measurements()
+    noisy_circuit.add(BasisMeasure(shots=shots))
+    mpqp_counts = _run_single(noisy_circuit, device, {}).counts
+
+    return float(jensenshannon(mpqp_counts, noisy_probs * sum(mpqp_counts)))
+
+
+def validate_noisy_circuit(
+    circuit: QCircuit,
+    shots: int = 1024,
+    device: AvailableDevice = AWSDevice.BRAKET_LOCAL_SIMULATOR,
+) -> bool:
+    """Validates our noise pipeline for a circuit.
+
+    Args:
+        circuit: The circuit (with potential noises).
+        shots: Number of shots in the basis measurement.
+        device: The device to be tested.
+
+    Returns:
+        Weather our noise pipeline matches the theory or not.
+    """
+    return exp_id_dist(circuit, shots, device) <= trust_int(circuit)
+
+
+def exp_id_dist_excess(circuit: QCircuit, shots: int = 1024) -> float:
+    """Computes the gap between theory and our noise pipeline for a circuit.
+
+    Args:
+        circuit: The circuit (with potential noises).
+        shots: Number of shots in the basis measurement.
+
+    Returns:
+        The gap between:
+
+        1. the distance between theory and our results;
+        2. and the trust interval.
+    """
+    return exp_id_dist(circuit, shots) - trust_int(circuit)
```

Comment:
Should we return the absolute value of this difference?

Reply:
indeed, but overall I think we'll have to redo a pass on this trust interval
thing, at least to prove it formally